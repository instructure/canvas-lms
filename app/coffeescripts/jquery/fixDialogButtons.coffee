define [
  'jquery'
  'underscore'
  'compiled/fn/preventDefault'
  'jqueryui/dialog'
], ($, _, preventDefault) ->

  $.fn.fixDialogButtons = ->
    this.each ->
      $dialog = $(this)
      $buttons = $dialog.find(".button-container:last .btn, button[type=submit]")
      if $buttons.length
        $dialog.find(".button-container:last, button[type=submit]").hide()
        buttons = $.map $buttons.toArray(), (button) ->
          $button = $(button)
          classes = $button.attr('class') ? ''
          id = $button.attr('id')

          # if you add the class 'dialog_closer' to any of the buttons,
          # clicking it will cause the dialog to close
          if $button.is('.dialog_closer')
            $button.off '.fixdialogbuttons'
            $button.on 'click.fixdialogbuttons', preventDefault -> $dialog.dialog('close')

          # make it so if you hit enter in the dialog, you submit the form
          if $button.prop('type') is 'submit' && $button[0].form
            classes += ' button_type_submit'
            $dialog.off '.fixdialogbuttons'
            $dialog.on 'keyup.fixdialogbuttons', (e) ->
              return unless $(e.target).filter('input:text').length
              $($button[0].form).submit() if e.keyCode is $.ui.keyCode.ENTER

          return {
            text: $button.text()
            "data-text-while-loading": $button.data("textWhileLoading")
            click: -> $button.click()
            class: classes
            id: id
          }
        # put the primary button(s) on the far right
        buttons = _.sortBy buttons, (button) ->
          if button.class.match(/btn-primary/) then 1 else 0
        $dialog.dialog "option", "buttons", buttons
