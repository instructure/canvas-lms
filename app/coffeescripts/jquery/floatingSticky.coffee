define [
  'underscore'
  'jquery'
], (_, $) ->

  # Floating sticky
  #
  # allows an element to float (using fixed positioning) as the user
  # scrolls, "sticking" to the top of the window. the difference from
  # a regular sticky implementation is that the element is constrained
  # by a containing element (or top and bottom elements), allowing the
  # element to float and stick, but only within the given bounds.
  #
  # to use, simply call .floatingSticky(containing_element) on a
  # jQuery object. optionally the top or bottom constraining element
  # can be overridden by providing {top:...} or {bottom:...} as the
  # last argument when calling .floatingSticky(...).
  #
  # the returned array has a floating sticky instance for each object
  # in the jQuery set, allowing calls to reposition() (in case the
  # element should be repositioned outside of a scroll/resize event)
  # or remove() to remove the floating sticky instance from the
  # element.

  instanceID = 0
  class FloatingSticky
    constructor: (el, container, options={}) ->
      @instanceID = "floatingSticky#{instanceID++}"

      @$window = $(window)
      @$el = $(el)
      @$top = $(options.top || container)
      @$bottom = $(options.bottom || container)

      @$el.data('floatingSticky', this)

      @$window.on "scroll.#{@instanceID} resize.#{@instanceID}", =>
        @reposition()
      @reposition()

    remove: ->
      @$window.off @instanceID
      @$el.data('floatingSticky', null)

    reposition: ->
      windowTop = @$window.scrollTop()
      windowHeight = @$window.height()

      # handle overscroll (up or down)
      if windowTop < 0
        windowTop = 0
      else
        windowTop = Math.min(windowTop, document.body.scrollHeight - windowHeight)

      # handle top of container
      containerTop = @$top.offset().top
      if windowTop < containerTop
        if windowTop == 0
          newTop = containerTop
        else
          newTop = containerTop - windowTop

      # handle bottom of container
      else
        newTop = 0
        elHeight = @$el.height()
        containerBottom = @$bottom.offset().top + @$bottom.height()

        # stay within the container
        if windowTop + elHeight > containerBottom
          newTop = containerBottom - elHeight - windowTop

        # but don't go above the container
        if newTop < containerTop - windowTop
          newTop = containerTop - windowTop

      @$el.css(top: newTop)

  $.fn.floatingSticky = (container, options={}) ->
    @map ->
      floatingSticky = $(this).data('floatingSticky')
      floatingSticky = new FloatingSticky(this, container, options) unless floatingSticky
      floatingSticky

  FloatingSticky
