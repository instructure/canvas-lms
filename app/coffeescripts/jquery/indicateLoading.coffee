#
# Copyright (C) 2012 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

define [
  'jquery'
], ($) ->
  
  # possible values for position are 'center' and 'after', see g_util_misc.scss
  # passign a position is optional and if ommited will use 'center'
  $.fn.indicateLoading = (position, deferred) ->
    unless deferred?
      deferred = position
      position = 'center'
    @each ->
      $this = $(this).addClass 'loading ' + position
      $.when(deferred).done ->
        $this.removeClass 'loading ' + position