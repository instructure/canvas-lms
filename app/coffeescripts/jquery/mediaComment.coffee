#mediaComment.coffee
define [
  'i18n!media_comments'
  'underscore'
  'vendor/jquery.ba-tinypubsub'
  'vendor/mediaelement-and-player'
  'jquery'
  'compiled/util/kalturaAnalytics'
], (I18n, _, pubsub, mejs, $, kalturaAnalytics) ->

  VIDEO_WIDTH = 550
  VIDEO_HEIGHT = 448
  $.extend mejs.MediaElementDefaults,
    # shows debug errors on screen
    # enablePluginDebug: false
    # path to Flash and Silverlight plugins
    pluginPath: '/images/mediaelement/'
    # default if the <video width> is not specified
    defaultVideoWidth: VIDEO_WIDTH
    # default if the <video height> is not specified
    defaultVideoHeight: VIDEO_HEIGHT

  unless INST.enableHtml5FirstVideos
    # prefer flash player, as it works more consistently
    # for now, but allow fallback to html5 (like on mobile)
    mejs.MepDefaults.mode = 'auto_plugin'

  mejs.MepDefaults.success = (mediaElement, domObject) ->
    kalturaAnalytics(this.mediaCommentId, mediaElement, INST.kalturaSettings)
    mediaElement.play()

  # track events in google analytics
  mejs.MepDefaults.features.push('googleanalytics')

  # enable the playback speed selector
  positionAfterSubtitleSelector = mejs.MepDefaults.features.indexOf('tracks') + 1
  mejs.MepDefaults.features.splice(positionAfterSubtitleSelector, 0, 'speed')

  getSourcesAndTracks = (id) ->
    dfd = new $.Deferred
    $.getJSON "/media_objects/#{id}/info", (data) ->
      # this 'when ...' is because right now in canvas, none of the mp3 urls actually work.
      # see: CNVS-12998
      sources = for source in data.media_sources when source.content_type isnt 'audio/mp3'
        "<source type='#{source.content_type}' src='#{source.url}' />"

      tracks = _.map data.media_tracks, (track) ->
        languageName = mejs.language.codes[track.locale] || track.locale
        "<track kind='#{track.kind}' label='#{languageName}' src='#{track.url}' srclang='#{track.locale}' />"

      types = _.map data.media_sources, (source) -> source.content_type
      dfd.resolve {sources, tracks, types, can_add_captions: data.can_add_captions}
    dfd

  createMediaTag = ({sourcesAndTracks, mediaType, height, width, mediaPlayerOptions}) ->
    tagType = if mediaType is 'video' then 'video' else 'audio'
    st_tags = sourcesAndTracks.sources.concat(sourcesAndTracks.tracks).join('')

    getElement = (tagType) ->
      $("<#{tagType} #{if mediaType is 'video' then "width='#{width}' height='#{height}'" else ''} controls>#{st_tags}</#{tagType}>")

    willPlayAudioInFlash = ->
      opts = $.extend({mode: 'auto'}, mejs.MediaElementDefaults, mejs.MepDefaults,mediaPlayerOptions)
      playback = mejs.HtmlMediaElementShim.determinePlayback(getElement('audio')[0], opts, mejs.MediaFeatures.supportsMediaTag, !!'isMediaTag', null)
      return playback.method != 'native'

    # We only need to do this if we try to play audio in a flash player.
    # A lot of our recorded audio is actually served up via video/mp4 or video/flv.
    # We need to trick the flash player into playing the video, but looking like
    # an audio player.
    if mediaType is 'audio' and sourcesAndTracks.types[0].match(/^video\//) and willPlayAudioInFlash()
      tagType = 'video'
      mediaPlayerOptions.mode = 'auto_plugin'
      mediaPlayerOptions.isVideo = false
      mediaPlayerOptions.videoHeight = height = 30

    getElement(tagType)

  mediaCommentActions =
    create: (mediaType, callback, onClose, defaultTitle) ->
      $("#media_recorder_container").removeAttr('id')
      this.attr('id', 'media_recorder_container')
      pubsub.unsubscribe 'media_comment_created'
      pubsub.subscribe 'media_comment_created', (data) =>
        callback(data.id, data.mediaType)

      initOpts = {modal: false, defaultTitle}
      initOpts.close = onClose.bind(this) if $.isFunction(onClose)

      $.mediaComment.init(mediaType, initOpts)

    show_inline: (id, mediaType = 'video', downloadUrl) ->
      # todo: replace .andSelf with .addBack when JQuery is upgraded.
      $holder = $(this).closest('.instructure_file_link_holder').andSelf().first()
      $holder.text I18n.t('loading', 'Loading media...')

      showInline = (id) ->
        width = Math.min ($holder.closest("div,p,table").width() || VIDEO_WIDTH), VIDEO_WIDTH
        height = Math.round width / 336 * 240
        getSourcesAndTracks(id).done (sourcesAndTracks) ->
          if sourcesAndTracks.sources.length
            mediaPlayerOptions =
               can_add_captions: sourcesAndTracks.can_add_captions
               mediaCommentId: id
               googleAnalyticsTitle: id

            $mediaTag = createMediaTag({sourcesAndTracks, mediaPlayerOptions, mediaType, height, width})
            $mediaTag.appendTo($holder.html(''))
            player = new MediaElementPlayer $mediaTag, mediaPlayerOptions
            $mediaTag.data('mediaelementplayer', player)
          else
            $holder.text I18n.t('media_still_converting', 'Media is currently being converted, please try again in a little bit.')

      if id is 'maybe'
        detailsUrl = downloadUrl.replace /\/download.*/, ""
        onError = ->
          $holder.text I18n.t 'messages.file_failed_to_load', "This media file failed to load"
        onSuccess = (data) ->
          if data.attachment?.media_entry_id isnt 'maybe'
            $holder.text ''
            showInline data.attachment.media_entry_id
          else
            onError()
        $.ajaxJSON detailsUrl, 'GET', {}, onSuccess, onError
      else
        showInline(id)

    show: (id, mediaType) ->
      $this = $(this)
      if dialog = $this.data('media_comment_dialog')
        dialog.dialog('open')
      else
        # Create a dialog box
        spaceNeededForControls = 35
        mediaType = mediaType || 'video'
        height = if mediaType is'video' then 426 else 180
        width = if mediaType is 'video' then VIDEO_WIDTH else 400
        $dialog = $('<div style="overflow: hidden; padding: 0;" />')
        $dialog.css('padding-top', '120px') if mediaType is 'audio'
        $dialog.dialog
          title: I18n.t('titles.play_comment', "Play Media Comment")
          width: width
          height: height
          modal: false
          resizable: false
          close: -> $this.data('mediaelementplayer').pause()

        # Populate dialog box with a video
        $dialog.disableWhileLoading getSourcesAndTracks(id).done (sourcesAndTracks) ->
          if sourcesAndTracks.sources.length
            mediaPlayerOptions = 
              can_add_captions: sourcesAndTracks.can_add_captions
              mediaCommentId: id
              googleAnalyticsTitle: id

            $mediaTag = createMediaTag({sourcesAndTracks, mediaPlayerOptions, mediaType, height: height-spaceNeededForControls, width})
            $mediaTag.appendTo($dialog.html(''))

            $this.data
              mediaelementplayer: new MediaElementPlayer $mediaTag, mediaPlayerOptions
              media_comment_dialog: $dialog
          else
            $dialog.text I18n.t('media_still_converting', 'Media is currently being converted, please try again in a little bit.')

  $.fn.mediaComment = (command, restArgs...) ->
    return console.log('Kaltura has not been enabled for this account') unless INST.kalturaSettings
    mediaCommentActions[command].apply(this, restArgs)
    this
