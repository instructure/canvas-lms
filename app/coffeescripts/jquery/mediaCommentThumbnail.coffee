define [
  'i18n!media_comments'
  'underscore'
  'str/htmlEscape'
  'jquery'
], (I18n, _, htmlEscape, $) ->

  MEDIA_COMMENT_THUMBNAIL_SIZES =
    normal: {width: 140, height: 100}
    small: {width: 70, height: 50}

  createMediaCommentThumbnail = (elem, size, keepOriginalText) ->
    return console.log('Kaltura has not been enabled for this account') unless INST.kalturaSettings

    $link = $(elem)
    dimensions = MEDIA_COMMENT_THUMBNAIL_SIZES[size] ? MEDIA_COMMENT_THUMBNAIL_SIZES.normal
    id = $link.data('media_comment_id') ||
         $.trim($link.find(".media_comment_id:first").text()) ||
         ((idAttr = $link.attr('id')) && idAttr.match(/^media_comment_/) && idAttr.substring(14)) ||
         $.trim($link.parent().find(".media_comment_id:first").text())

    if id
      domain = "https://#{INST.kalturaSettings.resource_domain}"

      backgroundUrl = "#{domain}/p/#{INST.kalturaSettings.partner_id}/thumbnail/entry_id/#{id}/width/" +
                      "#{dimensions.width}/height/#{dimensions.height}/bgcolor/000000/type/2/vid_sec/5"

      $thumbnail = $ "<span
                        style='background-image: url(#{htmlEscape(backgroundUrl)});'
                        class='media_comment_thumbnail media_comment_thumbnail-#{htmlEscape(size)}'
                      >
                        <span class='media_comment_thumbnail_play_button'>
                          #{htmlEscape I18n.t 'click_to_view', 'Click to view'}
                        </span>
                      </span>"

      $a = $link
      if keepOriginalText
        $a = $link.clone().empty().removeClass('instructure_file_link')
        $holder = $link.parent(".instructure_file_link_holder")
        if $holder.length
          $a.appendTo($holder)
        else
          $link.after($a)
      else
        $link.empty()

      $a
        .addClass('instructure_inline_media_comment')
        .append($thumbnail)
        .css
          backgroundImage: ''
          padding: 0

  # public API
  $.fn.mediaCommentThumbnail = (size='normal', keepOriginalText) ->
    this.each ->
      # defer each thumbnail generation till the next time through the event loop to not kill browser rendering,
      # has the effect of saying "only work on thumbnailing these while the browser is not doing something else"
      _.defer(createMediaCommentThumbnail, this, size, keepOriginalText)
