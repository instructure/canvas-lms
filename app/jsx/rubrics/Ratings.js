/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import _ from 'lodash'
import $ from 'jquery'
import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import Text from '@instructure/ui-elements/lib/components/Text'
import I18n from 'i18n!edit_rubric'

import { ratingShape, tierShape } from './types'

const pointString = (points, endOfRangePoints) => {
  if (endOfRangePoints !== null) {
    return I18n.t('%{points} to >%{endOfRangePoints} pts', {
      points: I18n.toNumber(points, { precision : 1 }),
      endOfRangePoints: I18n.toNumber(endOfRangePoints, { precision : 1 })
    })
  }
  else {
    return I18n.t('%{points} pts', {
      points: I18n.toNumber(points, { precision : 1 })
    })
  }
}

export const Rating = (props) => {
  const {
    assessing,
    classes,
    description,
    endOfRangePoints,
    footer,
    long_description,
    points,
    onClick,
    tierColor,
    hidePoints,
    isSummary,
    selected
  } = props

  const shaderStyle = { backgroundColor: tierColor }
  const triangleStyle = { borderBottomColor: tierColor }

  const ratingPoints = () => (
    <div className="rating-points">
      <Text size="small" weight="bold">
        {pointString(points, endOfRangePoints)}
      </Text>
    </div>
  )

  return (
    <div
      className={classes}
      onClick={assessing ? onClick : null}
      onKeyPress={(e) => e.key === 'Enter' ? onClick() : null}
      role={assessing ? "button" : null}
      tabIndex={assessing ? 0 : null}
    >
      {hidePoints ? null : ratingPoints()}
      <div className="rating-description">
        <Text size="small" lineHeight="condensed" weight="bold">
          {description}
        </Text>
      </div>
      <Text size="small" lineHeight="condensed">
        {long_description}
      </Text>
      {
        footer !== null ? (
          <div className="rating-footer">
            {footer}
          </div>
        ) : null
      }
      <div className='shader' style={shaderStyle}
        aria-label={isSummary || !selected ? null : I18n.t('This rating is selected')}>
        <div className="triangle" style={triangleStyle}/>
      </div>
    </div>
  )
}

const getDefaultColor = (points, defaultMasteryThreshold) => {
  if (points >= defaultMasteryThreshold) {
    return '#8aac53'
  } else if (points >= defaultMasteryThreshold/2) {
    return '#e0d773'
  } else {
    return '#df5b59'
  }
}

const getCustomColor = (points, pointsPossible, customRatings) => {
  const sortedRatings = _.sortBy(customRatings, 'points').reverse()
  const scaledPoints = pointsPossible > 0 ? points * (sortedRatings[0].points / pointsPossible) : points
  const selectedRating = _.find(sortedRatings, (rating) => ( scaledPoints >= rating.points ))
  if (selectedRating) {
    return `#${selectedRating.color}`
  } else {
    return `#${_.last(sortedRatings).color}`
  }
}

Rating.propTypes = {
  ...tierShape,
  assessing: PropTypes.bool.isRequired,
  footer: PropTypes.node,
  selected: PropTypes.bool,
  hidePoints: PropTypes.bool,
  isSummary: PropTypes.bool.isRequired,
}
Rating.defaultProps = {
  footer: null,
  selected: false,
  endOfRangePoints: null, // eslint-disable-line react/default-props-match-prop-types
  hidePoints: false
}

const Ratings = (props) => {
  const {
    assessing,
    customRatings,
    defaultMasteryThreshold,
    footer,
    tiers,
    points,
    pointsPossible,
    hidePoints,
    onPointChange,
    isSummary,
    useRange
  } = props

  const pairs = tiers.map((tier, index) => {
    const next = tiers[index + 1]
    return { current: tier.points, next: next ? next.points : null }
  })

  const currentIndex = () => pairs.findIndex(({ current, next }) => {
    const currentMatch = points === current
    const withinRange = points > next && points <= current
    const zeroAndInLastRange = points === 0 && next === null
    return currentMatch || (useRange && (withinRange || zeroAndInLastRange))
  })

  const getRangePoints = (currentPoints, nextTier) => {
    if (nextTier) {
      return currentPoints === nextTier.points ? null : nextTier.points
    } else if (currentPoints !== 0) {
      return 0
    }
    return null
  }

  const getTierColor = (selected) => {
    if (!selected) { return 'transparent' }
    if (customRatings && customRatings.length > 0) {
      return getCustomColor(points, pointsPossible, customRatings)
    } else {
      return getDefaultColor(points, defaultMasteryThreshold)
    }
  }

  const handleClick = (tierPoints) => {
    onPointChange(tierPoints)
    $.screenReaderFlashMessage(I18n.t('Rating selected'))
  }

  const selectedIndex = points !== undefined ? currentIndex() : null
  const ratings = tiers.map((tier, index) => {
    const selected = selectedIndex === index
    if (isSummary && !selected) return null
    const classes = classNames({
      'rating-tier': true,
      'selected': selected,
    })

    return (
      <Rating
        key={index} // eslint-disable-line react/no-array-index-key
        assessing={assessing}
        classes={classes}
        endOfRangePoints={useRange ? getRangePoints(tier.points, tiers[index + 1]) : null}
        footer={footer}
        onClick={() => handleClick(tier.points)}
        tierColor={getTierColor(selected)}
        hidePoints={isSummary || hidePoints}
        isSummary={isSummary}
        selected={selected}
        {...tier}
      />
    )
  }).filter((v) => v !== null)

  const defaultRating = () => (
    <Rating
      key={0}
      classes="rating-tier"
      description={I18n.t('No details')}
      footer={footer}
      points={0}
      hidePoints={isSummary || hidePoints}
    />
  )

  return (
    <div className={classNames("rating-tier-list", { 'react-assessing': assessing })}>
      {ratings.length > 0 || !isSummary ? ratings : defaultRating()}
    </div>
  )
}
Ratings.propTypes = {
  ...ratingShape,
  assessing: PropTypes.bool.isRequired,
  footer: PropTypes.node,
  onPointChange: PropTypes.func,
  isSummary: PropTypes.bool.isRequired,
  hidePoints: PropTypes.bool
}
Ratings.defaultProps = {
  footer: null,
  hidePoints: false,
  onPointChange: () => { }
}

export default Ratings
