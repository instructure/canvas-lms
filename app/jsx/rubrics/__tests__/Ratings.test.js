/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import $ from 'jquery'
import { shallow } from 'enzyme'
import sinon from 'sinon'
import Ratings, { Rating } from '../Ratings'

// This is needed for $.screenReaderFlashMessageExclusive to work.
import 'compiled/jquery.rails_flash_notifications' // eslint-disable-line

describe('The Ratings component', () => {
  const props = {
    assessing: false,
    footer: null,
    tiers: [
      { description: 'Superb', points: 10 },
      { description: 'Meh', long_description: 'More Verbosity', points: 5 },
      { description: 'Subpar', points: 1 }
    ],
    defaultMasteryThreshold: 10,
    points: 5,
    pointsPossible: 10,
    isSummary: false,
    useRange: false
  }

  const component = (mods) => shallow(<Ratings {...{ ...props, ...mods }} />)
  it('renders the root component as expected', () => {
    expect(component().debug()).toMatchSnapshot()
  })

  it('renders the Rating sub-components as expected when range rating enabled', () => {
    const useRange = true
    component({ useRange }).find('Rating')
      .forEach((el) => expect(el.shallow().debug()).toMatchSnapshot())
  })

  it('highlights the right rating', () => {
    const ratings = (points, useRange = false) =>
      component({ points, useRange }).find('Rating').map((el) => el.shallow().hasClass('selected'))

    expect(ratings(10)).toEqual([true, false, false])
    expect(ratings(8)).toEqual([false, false, false])
    expect(ratings(8, true)).toEqual([true, false, false])
    expect(ratings(5)).toEqual([false, true, false])
    expect(ratings(3)).toEqual([false, false, false])
    expect(ratings(3, true)).toEqual([false, true, false])
    expect(ratings(1)).toEqual([false, false, true])
    expect(ratings(0, true)).toEqual([false, false, true])
    expect(ratings(undefined)).toEqual([false, false, false])
  })

  it('calls onPointChange and flashes VO message when a rating is clicked', () => {
    const onPointChange = sinon.spy()
    const flashMock = jest.spyOn($, 'screenReaderFlashMessage')
    const el = component({ onPointChange })

    el.find('Rating').first().prop('onClick').call()
    expect(onPointChange.args[0]).toEqual([10])
    expect(flashMock).toHaveBeenCalledTimes(1)
    flashMock.mockRestore()
  })

  it('uses the right default mastery level colors', () => {
    const mastery = (points) =>
      component({ points }).find('Rating').map((el) => el.prop('tierColor'))
    expect(mastery(10)).toEqual(["#8aac53", 'transparent', 'transparent'])
    expect(mastery(5)).toEqual(['transparent', "#e0d773", 'transparent'])
    expect(mastery(1)).toEqual(['transparent', 'transparent', "#df5b59"])
  })

  it('uses the right custom rating colors', () => {
    const customRatings = [
      {points: 10, color: "09BCD3"},
      {points: 5, color: "65499D"},
      {points: 1, color: "F8971C"}
    ]
    const ratings = (points, useRange = false) =>
      component({points, useRange, customRatings}).find('Rating').map((el) => el.prop('tierColor'))
    expect(ratings(10)).toEqual(['#09BCD3', 'transparent', 'transparent'])
    expect(ratings(5)).toEqual(['transparent', '#65499D', 'transparent'])
    expect(ratings(1)).toEqual(['transparent', 'transparent', '#F8971C'])
    expect(ratings(0, true)).toEqual(['transparent', 'transparent', '#F8971C'])
  })

  describe('custom ratings', () => {
    const customRatings = [
      {points: 100, color: "100100"},
      {points: 60, color: "606060"},
      {points: 10, color: "101010"},
      {points: 1, color: "111111"}
    ]
    const ratings = (points, pointsPossible = 10) =>
      component({ points, pointsPossible, customRatings, useRange: true }).find('Rating').map((el) => el.prop('tierColor'))

    it('scales points to custom ratings', () => {
      expect(ratings(10)).toEqual(['#100100', 'transparent', 'transparent'])
      expect(ratings(6)).toEqual(['#606060', 'transparent', 'transparent'])
      expect(ratings(5)).toEqual(['transparent', '#101010', 'transparent'])
      expect(ratings(4.4)).toEqual(['transparent', '#101010', 'transparent'])
      expect(ratings(1)).toEqual(['transparent', 'transparent', '#101010'])
      expect(ratings(0.1)).toEqual(['transparent', 'transparent', '#111111'])
      expect(ratings(0)).toEqual(['transparent', 'transparent', '#111111'])
    })

    it('does not scale points if pointsPossible is 0', () => {
      expect(ratings(10, 0)).toEqual(['#101010', 'transparent', 'transparent'])
      expect(ratings(4, 0)).toEqual(['transparent', '#111111', 'transparent'])
    })
  })

  it('does not scale points if points possible is 0', () => {

  })

  it('is navigable and clickable when assessing', () => {
    const onClick = sinon.spy()
    const wrapper = shallow(<Rating {...props.tiers[0]} assessing onClick={onClick} />)
    const div = wrapper.find('div').at(0)
    expect(div.prop('tabIndex')).toEqual(0)
    div.simulate('click')
    expect(onClick.called).toBe(true)
  })

  it('is not navigable or clickable when not assessing', () => {
    const onClick = sinon.spy()
    const wrapper = shallow(<Rating {...props.tiers[0]} assessing={false} onClick={onClick} />)
    const div = wrapper.find('div').at(0)
    expect(div.prop('tabIndex')).toBeNull()
    expect(div.prop('role')).toBeNull()
    div.simulate('click')
    expect(onClick.called).toBe(false)
  })

  it('only renders the single selected Rating with a footer in summary mode', () => {
    const el = component({ points: 5, isSummary: true, footer: <div>ow my foot</div> })
    const ratings = el.find('Rating')

    expect(ratings).toHaveLength(1)

    const rating = ratings.at(0)
    expect(rating.shallow().debug()).toMatchSnapshot()
  })

  it('renders a default rating if none of the ratings are selected', () => {
    const el = component({ points: 6, isSummary: true, footer: <div>ow my foot</div> })
    const ratings = el.find('Rating')

    expect(ratings).toHaveLength(1)

    const rating = ratings.at(0)
    expect(rating.shallow().debug()).toMatchSnapshot()
  })

  it('hides points on the default rating if points are hidden', () => {
    const el = component({ points: 6, isSummary: true, footer: <div>ow my foot</div> })
    const rating = el.find('Rating')
    expect(rating.prop('hidePoints')).toBe(true)
  })
})
