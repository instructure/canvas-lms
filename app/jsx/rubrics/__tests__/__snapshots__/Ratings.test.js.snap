// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Ratings component only renders the single selected Rating with a footer in summary mode 1`] = `
"<div className=\\"rating-tier selected\\" onClick={{...}} onKeyPress={[Function]} role={{...}} tabIndex={{...}}>
  <div className=\\"rating-description\\">
    <Text size=\\"small\\" lineHeight=\\"condensed\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
      Meh
    </Text>
  </div>
  <Text size=\\"small\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\">
    More Verbosity
  </Text>
  <div className=\\"rating-footer\\">
    <div>
      ow my foot
    </div>
  </div>
  <div className=\\"shader\\" style={{...}} aria-label={{...}}>
    <div className=\\"triangle\\" style={{...}} />
  </div>
</div>"
`;

exports[`The Ratings component renders a default rating if none of the ratings are selected 1`] = `
"<div className=\\"rating-tier\\" onClick={{...}} onKeyPress={[Function]} role={{...}} tabIndex={{...}}>
  <div className=\\"rating-description\\">
    <Text size=\\"small\\" lineHeight=\\"condensed\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
      No details
    </Text>
  </div>
  <Text size=\\"small\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\" />
  <div className=\\"rating-footer\\">
    <div>
      ow my foot
    </div>
  </div>
  <div className=\\"shader\\" style={{...}} aria-label={{...}}>
    <div className=\\"triangle\\" style={{...}} />
  </div>
</div>"
`;

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 1`] = `
"<div className=\\"rating-tier\\" onClick={{...}} onKeyPress={[Function]} role={{...}} tabIndex={{...}}>
  <div className=\\"rating-points\\">
    <Text size=\\"small\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
      10.0 to &gt;5.0 pts
    </Text>
  </div>
  <div className=\\"rating-description\\">
    <Text size=\\"small\\" lineHeight=\\"condensed\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
      Superb
    </Text>
  </div>
  <Text size=\\"small\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\" />
  <div className=\\"shader\\" style={{...}} aria-label={{...}}>
    <div className=\\"triangle\\" style={{...}} />
  </div>
</div>"
`;

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 2`] = `
"<div className=\\"rating-tier selected\\" onClick={{...}} onKeyPress={[Function]} role={{...}} tabIndex={{...}}>
  <div className=\\"rating-points\\">
    <Text size=\\"small\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
      5.0 to &gt;1.0 pts
    </Text>
  </div>
  <div className=\\"rating-description\\">
    <Text size=\\"small\\" lineHeight=\\"condensed\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
      Meh
    </Text>
  </div>
  <Text size=\\"small\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\">
    More Verbosity
  </Text>
  <div className=\\"shader\\" style={{...}} aria-label=\\"This rating is selected\\">
    <div className=\\"triangle\\" style={{...}} />
  </div>
</div>"
`;

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 3`] = `
"<div className=\\"rating-tier\\" onClick={{...}} onKeyPress={[Function]} role={{...}} tabIndex={{...}}>
  <div className=\\"rating-points\\">
    <Text size=\\"small\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
      1.0 to &gt;0.0 pts
    </Text>
  </div>
  <div className=\\"rating-description\\">
    <Text size=\\"small\\" lineHeight=\\"condensed\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
      Subpar
    </Text>
  </div>
  <Text size=\\"small\\" lineHeight=\\"condensed\\" as=\\"span\\" letterSpacing=\\"normal\\" />
  <div className=\\"shader\\" style={{...}} aria-label={{...}}>
    <div className=\\"triangle\\" style={{...}} />
  </div>
</div>"
`;

exports[`The Ratings component renders the root component as expected 1`] = `
"<div className=\\"rating-tier-list\\">
  <Rating assessing={false} classes=\\"rating-tier\\" endOfRangePoints={{...}} footer={{...}} onClick={[Function]} tierColor=\\"transparent\\" hidePoints={false} isSummary={false} selected={false} description=\\"Superb\\" points={10} />
  <Rating assessing={false} classes=\\"rating-tier selected\\" endOfRangePoints={{...}} footer={{...}} onClick={[Function]} tierColor=\\"#e0d773\\" hidePoints={false} isSummary={false} selected={true} description=\\"Meh\\" long_description=\\"More Verbosity\\" points={5} />
  <Rating assessing={false} classes=\\"rating-tier\\" endOfRangePoints={{...}} footer={{...}} onClick={[Function]} tierColor=\\"transparent\\" hidePoints={false} isSummary={false} selected={false} description=\\"Subpar\\" points={1} />
</div>"
`;
